### PR TITLE
feat: add validations to census builder

### LINKED ISSUE #35: cell_census_builder: validate no dups
Add validations that there are no duplicate indices in all Measurement Layers, for all Experiments:
* X[raw] - all (soma_dim_0, soma_dim_1) pairs are unique
* obs/var - all soma_joinid are unique
Then configure `allow_duplicates=True` (to improve TileDB performance). This requires parameterizing the `allow_duplicates` attribute via `platform_config` in TileDB-SOMA library.

The removal of `allow_duplicates` on the underlying array schemas provides a significant read performance improvement.

### PR DESCRIPTION
**Pull request by atolopko-czi**
Resolves #35 

- no dups in X matrix (unique coords)
- no dups in obs or var soma_joins

Testing in progress, but early review may be useful.

### CODE REVIEW

File: tools/cell_census_builder/__main__.py
Comment (atolopko-czi) [inline tools/cell_census_builder/__main__.py:126]: Fail here instead of later with an arcane, harder-to-troubleshoot error in code that doesn't expect empty set of datasets. Useful when using `--test-first-n`, which may result in 0 admissible datasets.

File: tools/cell_census_builder/validate.py
Comment (atolopko-czi) [inline tools/cell_census_builder/validate.py:321]: As it turns out, `_validate_X_layers_contents` catches the dup coords as well (e.g. `AssertionError: homo_sapiens:4 'raw' nnz mismatch 3618682 vs 1809341`, on a test corpus that duplicates every obs row). So this check might not be strictly necessary, though it does make the semantics of the check explicit.

**Review by bkmartinjr — COMMENTED** [review]
The motivation for this change was to enable use of the TileDB schema `allow_duplicates` feature (for the X sparse matrices).  This schema config (allowing dups at the core DB level) unlocks additional read performance, as the core DB does not have to perform duplicate consolidation upon read.

Is the plan to do that in another PR?  Without this TileDB schema change, we essentially are doing the check twice (once in the core DB because `allow_updates=False`, and once in our own code).

**Comment by atolopko-czi** [comment]
> The motivation for this change was to enable use of the TileDB schema `allow_duplicates` feature (for the X sparse matrices). This schema config (allowing dups at the core DB level) unlocks additional read performance, as the core DB does not have to perform duplicate consolidation upon read.
> 
> Is the plan to do that in another PR? Without this TileDB schema change, we essentially are doing the check twice (once in the core DB because `allow_updates=False`, and once in our own code).

I assumed (incorrectly) that the `allow_dups` setting change was separately planned work. Will take that on as part of the story.

**Comment by bkmartinjr** [comment]
> I assumed (incorrectly) that the allow_dups setting change was separately planned work.

it is fine either way with me.  I'll approve the PR, and you can either land, or hit me up for a second review!

**Review by bkmartinjr — APPROVED** [review]
approved assuming enabling allow_dups in a second PR.  If not, I'm happy to re-review!  TY.

**Comment by atolopko-czi** [comment]
Parent issue replaced by https://github.com/chanzuckerberg/cell-census/issues/61. Closing this PR.

We _could_ salvage two minor pieces from this PR:
* https://github.com/chanzuckerberg/cell-census/pull/59/files#diff-84c0d083cd4389aad3dcca3747de5b150f9e9f157074304057471f398df803c7R221-R231
* https://github.com/chanzuckerberg/cell-census/pull/59/files#diff-5eb2ff26a895b5d0440289ad915a57f054370888653a8ee168c34882ca2d4c7fR124-R126

**Pull request closed without merging**
